Open mail

Umar222 opened the pull request.

Umar222 committed:
Класс с эементами почты

Umar222 committed:
Класс с тестами почты

Umar222 committed:
1) Добавил локаторы полей логина/пароля, кнопки "Войти", 2) Объеденил локаторы в метод авторизации на почте

Umar222 committed:
Написал метод авторизации на почте

Umar222 committed:
1) удалил метод открытия сайт из конфигурационного файла, 2) Добавил метод вызыва сайта yandex.ru

Umar222 committed:
Написал метод checkEnterMessage проверки входящих писем по теме Simbirsoft theme

Umar222 committed:
Добавил локатор темы сообщения messageTheme

Plodick commented on src/main/java/ru/Ilnur/Pages/mailPage.java:
```diff
@@ -0,0 +1,59 @@
+    public List<WebElement> messageTheme;
+
+
+
```
Постарайся избегать пустых строк лишних

Plodick commented on src/main/java/ru/Ilnur/Pages/mailPage.java:
```diff
@@ -0,0 +1,59 @@
+        loginField.submit();
+        passwordField.sendKeys(password);
+        submit.click();
+
```
Еще несколько.

Plodick commented on src/main/java/ru/Ilnur/Pages/mailPage.java:
```diff
@@ -0,0 +1,59 @@
+        openMailButton.click();
+    }
+
+    public void openMail(String mail, String password) {
```
Названия метода должно чуть более конкретно отражать то, что он делает. Например этот метод не просто открывает почту, а авторизуется.

Plodick commented on src/main/java/ru/Ilnur/Pages/mailPage.java:
```diff
@@ -0,0 +1,59 @@
+
+import java.util.List;
+
+public class mailPage {
```
Немного помучаю по стилю кода: названия классов с большой буквы.

Plodick commented on src/main/java/ru/Ilnur/Pages/mailPage.java:
```diff
@@ -0,0 +1,59 @@
+
+    public WebDriver driver;
+
+    public mailPage(WebDriver driver) {
```
Сначала поля, потом конструкторы, потом методы.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,60 @@
+    public void checkEnterMessage() {
+        ArrayList<String> tabs2 = new ArrayList<String>(driver.getWindowHandles());
+        driver.switchTo().window(tabs2.get(1));
+        List<WebElement> elements = driver.findElements(By.xpath("//span[@class=\"mail-MessageSnippet-Item mail-MessageSnippet-Item_subject\"]"));
```
Элементы и взаимодействия с ними необходимо инкапсулировать в Page Object'ах.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,60 @@
+            if (title.getText().equals("Simbirsoft theme"))
+                count++;
+        }
+        System.out.println("Писем по теме найдено " + count);
```
Спешу обрадовать, что данный тест ничего не проверяет - необходимо фиксировать значения писем перед отправкой сообщения - отправлять его и проверять с текущим.

Umar222 committed:
1) Переместил конструктор под поля 2) Добавил новый метод перехода к следующей вкладке swichNewPage 3) Добавил аннотации Step ко всем методам 4) Уменьшил количество отступов между локаторами и методами 5) Сделал название класса с заглавной буквы

Umar222 committed:
1) Создал метод метод перехода на следующую вкладку 2) Добавил проверку названия темы assertEquals(title.getText(), "Simbirsoft theme"); хотя мне кажется что подобные проверки не уместны для тестов т.к яндекс присылают рекламму, название темы может меняться и при прохождении ассерта тесты всегда будут падать

Plodick commented on src/main/java/ru/Ilnur/Pages/MailPage.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Step("Авторизация на почте")
+    public void openMail(String mail, String password) {
```
Этот метод в некоторой степени нарушает  принцип Page Object - так как данная логика относится к странице авторизации, а не почты.

Plodick replied:
Ты метод перенес - но старый не удалил

Plodick commented on src/main/java/ru/Ilnur/Pages/MailPage.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Step("Переход на новую вкладку")
+    public void switchNewPage(int pageNumber) {
```
А этот метод вообще ни к одной странице не относится, я бы вообще не заморачивался и поместил бы это туда, где это переключение и происходит.

Plodick replied:
Подчищай то, чем не пользуешься.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,81 @@
+    @Test
+    public void checkEnterMessage() {
+        mailPage.switchNewPage(1);
+        for (WebElement title : mailPage.messageTheme) {
```
Этот "тест" уже с этой строчки теряет смысл. Он проходит по всем сообщениям, а assert на первом же сообщении не удовлетворяющему твоему условию вылетит. Проверяем мы всё таки не чтобы все тесты в ящике были одной темы, а то, что новое письмо приходит.

Umar222 replied:
В условиях написано "Проверить, что с каждым запуском теста, кол-во писем **с темой Simbirsoft theme увеличивается.**"
Я не знаю как еще по другому проверить это условие

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,81 @@
+        mailPage.switchNewPage(1);
+        for (WebElement title : mailPage.messageTheme) {
+            assertEquals(title.getText(), "Simbirsoft theme");
+            if (title.getText().equals("Simbirsoft theme"))
```
Молчу уже про то, что вместо такого цикла, ты можешь просто сделать поиск по тексту сразу нужной коллекции, например, юзая xpath.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,81 @@
+            if (title.getText().equals("Simbirsoft theme"))
+                n++;
+        }
+        if (n < 1) {
```
А эти условия даже не тестовые. Тест должен что-то проверять и давать результат: работает данная функциональность или нет. А этот метод несет больше утилитарный смысл - вывести число сообщений.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Description("Проверка отображения нового письма")
+    @Test
+    public void checkAddingNewMessage() {
```
Ну и да, на этот момент тебе достаточно одного теста.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,81 @@
+        }
+        System.out.println("Первый счетчик " + n);
+        System.out.println("Второй счетчик " + count);
+        assert (count > n);
```
Советую юзать ассерты junit - они дают больше информации. И еще: логические ассерты оставляй на крайний случай.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,81 @@
+        mailPage.newMessage();
+        mailPage.addMailAddress("testovumar@yandex.ru", "Simbirsoft theme");
+        mailPage.sendMessage();
+        mailPage.modalTitle().isDisplayed();
```
Это строка ничего не делает - она возвращает булево значение.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,81 @@
+        mailPage.sendMessage();
+        mailPage.modalTitle().isDisplayed();
+        driver.navigate().refresh();
+        for (WebElement title : mailPage.messageTheme) {
```
Как я уже сказал, можно сделать поиск по тексту и еще, так как это логика страницы, надо вынести это в POM.

Umar222 committed:
1)Создал класс авторизации 2) Добавил локаторы loginField, passwordField, submit 2) Объеденил локаторы в метод authMail 3) В метод authMail добавил аннотацию @Step

Umar222 committed:
1) Добавил экземпляр класса MailAuthPage в метод setUp()
2) Метод switchNewPage вернул из класса MailPage
3) Поменял метод проверки отображения текста на assertEquals(mailPage.modalTitle().getText(), "Письмо отправлено");
4) Добавил отдельный метод пересчитывания писем на почте mailCount, добавил его в класс MailPage
5) Поменял метод сравнения количества писем на jUnit assert, assertTrue(count > n);

Plodick commented on src/main/java/ru/Ilnur/Pages/MailAuthPage.java:
```diff
@@ -0,0 +1,37 @@
+        passwordField.sendKeys(password);
+        submit.click();
+    }
+
```
Пустые строки

Plodick commented on src/main/java/ru/Ilnur/Pages/MailPage.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Step("Ввожу адрес и тему письма")
+    public void addMailAddress(String email, String theme) {
```
Я бы заменил на более говорящее что-то и общее: например, fillMessage

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,70 @@
+    static MailPage mailPage;
+    int n;
+
+    public void switchNewPage(int pageNumber) {
```
Метод сделай статическим и вынеси лучше в какой-нибудь класс хелперов/утилит.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,70 @@
+    public static WebDriver driver = new ChromeDriver();
+    static MailAuthPage mailAuthPage;
+    static MailPage mailPage;
+    int n;
```
А это нигде не используется.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,70 @@
+    public void checkAddingNewMessage() {
+
+        switchNewPage(1);
+        int n = mailPage.mailCount("Simbirsoft theme");
```
Однобуквенные переменные не принято использовать. У нас не математические задачи, да и обфускацию проводить не требуется.

Umar222 replied:
Тут я отталкивался от описания задачи, там написано было "n"

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,70 @@
+
+        int count = mailPage.mailCount("Simbirsoft theme");
+
+        assertTrue(count > n);
```
Добавь сообщение, которое будет выпадать при ошибке.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,70 @@
+        int count = mailPage.mailCount("Simbirsoft theme");
+
+        assertTrue(count > n);
+
```
Пустые строки, которое не несут смысл, допустим, в логическом разделении блоков теста, тоже лучше подчистить.

Umar222 committed:
1) Убрал лишние отсупы в методе checkAddingNewMessage 2) Добавил к ассертам сообщение в случае падения теста 3) Удалил переменную "n" 4) Переименовал переменные счетчиков на "count1" и "count2" 5) По молился что бы на этот раз все было хорошо

Umar222 committed:
1) Создал абстрактный класс для общих сущностей которые могут использоваться во всех классах

Umar222 committed:
1) Добавил экземпляр класса General 2) Метод switchNewPage объявил статическим, переместил в класс General, и вызвал в методе checkAddingNewMessage

Plodick commented on src/main/java/ru/Ilnur/Pages/General.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.util.ArrayList;
+
+public class General {
```
Тоже название не очень отражает содержимое, лучше используй клешированное BasePage

Plodick commented on src/main/java/ru/Ilnur/Pages/MailAuthPage.java:
```diff
@@ -0,0 +1,37 @@
+        submit.click();
+    }
+
+
```
Пустые строки все еще есть

Plodick commented on src/main/java/ru/Ilnur/Pages/MailPage.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Step("Ввожу адрес и тему письма")
+    public void addMailAddress(String email, String theme) {
```
Название

Umar222 replied:
Не понимаю что не так с названием? Вполне себе говорящее название метода

Plodick replied:
Название метода гласит - добавить адрес почты.
Какой почты? Куда добавить? У меня не вызывает ассоциаций в том, что именно этот метод заполняет поля темы и сообщения в письме.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,62 @@
+    public static WebDriver driver = new ChromeDriver();
+    static MailAuthPage mailAuthPage;
+    static MailPage mailPage;
+    static General general;
```
Зачем делать их статиками, если ты вызываешь конструкторами?

Umar222 replied:
потому что я их вызываю их в статическом методе setUp

Plodick replied:
Ну, сейчас можешь оставить, но если ты создашь тест с несколькими методами, которые должны работать независимо - данная реализация работать не будет, потому что состояние (в данном случае драйвер) надо будет переопределять.

Umar222 committed:
1) Перенес метод openMail в класс MailAuthPage 2) Добавил метод счета писем mailCount

Umar222 committed:
Переименовал класс

Umar222 committed:
1) Добавил аннотацию Step на методы Field и search 2) Перенес конструктор под локаторы

Plodick commented on src/main/java/ru/Ilnur/Pages/MailAuthPage.java:
```diff
@@ -0,0 +1,37 @@
+
+
+
+
```
Ну убери пустые строки

Umar222 committed:
Убрал лишние отступы

Umar222 committed:
Переименовал метод addMailAddress на inputMailAddress

Umar222 committed:
1)Удалил не используемые библиотеки 2) Удалил не нужные отступы

Umar222 committed:
Изменилось название метода addMailAddress на inputMailAddress

Plodick commented on src/main/java/ru/Ilnur/Pages/CalculatorPage.java:
```diff
@@ -7,64 +7,48 @@
-
     @FindBy(xpath = "//span[@jsname = \"VssY5c\"]")
     private WebElement numberField;
-
```
А зачем ты тут удалил пустые строки?

Plodick commented on src/main/java/ru/Ilnur/Pages/MailAuthPage.java:
```diff
@@ -0,0 +1,32 @@
+    @FindBy(xpath = "//input[@id = \"passp-field-login\"]")
+    private WebElement loginField;
+    @FindBy(xpath = "//input[@id = \"passp-field-passwd\"]")
+    private WebElement passwordField;
```
Между полями тоже отступ строка.

Plodick commented on src/test/java/ru/Ilnur/SimbirSoftTests/yandexMailTests.java:
```diff
@@ -0,0 +1,60 @@
+        assertEquals(mailPage.modalTitle().getText(), "Письмо отправлено", "Текст не отобразился");
+        driver.navigate().refresh();
+        int count2 = mailPage.mailCount("Simbirsoft theme");
+        assertTrue(count2 > count1, "Новое письмо не создалось");
```
Я бы всё таки уточнил в данном тесте, что придет именно одно письмо.  Здесь информативнее будет assertEquals(count1 + 1, count2).

Plodick replied:
Условие и проверку можешь сделать другую, но суть в том, что прийти должно только одно письмо.

Umar222 committed:
Добавил отступы между локаторами

Umar222 committed:
Добавил отступы между локаторами

Umar222 committed:
Добавил отступы между локаторами

Umar222 committed:
1)В методе checkAddingNewMessage поменял проверку assertTrue на assertEquals(count1 + 1, count2, "Новое письмо не создалось"); 2) Удалил лишние отступы 3) Удалил библиотеку assertTrue

Umar222 committed:
1) Убрал лишние отступы 2)Метод inputSearchField перенс в класс SearchPage

Umar222 committed:
1) Убрал лишние отступы 2) Добавил метод inputSearchField 3) Переместил конструктор под локаторы

Umar222 committed:
Сделал авто форматирование строк кода согласно общему стилю

Umar222 committed:
Добавил sourceCompatibility = 1.8 т.к была ошибка в тестах

Umar222 committed:
Написал тест на проверку корректный почты пользователя George, при проверке ориентировался что пользователю George присвоен id и по нему проверял совпадение почты

Umar222 committed:
Поменял проверку по id на проверку по параметру last_name
